### DISCUSSION TITLE
warning: loc idx out of lower bound

### DISCUSSION THREAD

**Discussion by Jimbo994** [deepmodeling/deepmd-kit#21] [Q&A] [answered]
Hi,
 
I am currently running a system containing water molecules and 2 Ruthenium atoms. The total system contains 193 atoms.
I believe I prepared all the data correctly. This is what type.raw looks like:

`2 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0`

This is what my input file looks like:

```
{
    "with_distrib":     false,
    "_comment": " model parameters",
    "use_smooth":       false,
    "sel_a":            [60, 30, 2],
    "sel_r":            [60, 30, 2],
    "rcut":             6.00,
    "axis_rule":        [0, 1, 0, 0, 0, 0,   0, 0, 0, 0, 0, 1,   0, 1, 0, 0, 1, 1],
    "_comment": " For Ru2Ru3 (Ru:2, O:1, H:0) ",
    "_comment": " This means first six numbers are for H, then six for O etc.",
    "_comment": " 6 integers should be provided. The first three for the first axis, while the last three for the second axis. ",
    "_comment": " the first one specifies if the axis atom is fully-local-coordinated (0) or radial-only-coordinated (1). ",
    "_comment": " The second integer specifies the type of the axis atom. If this number is less than 0, saying t < 0, then this axis exclude atom of type -(t+1) ",
    "_comment": " Third integer denotes the sth nearest neighbour ",
    "_comment": " default rule: []",
    "_comment": " user defined rule: for each type provides two axes, ",
    "_comment": "                    for each axis: (a_or_r, type, idx)",
    "_comment": "                    if type < 0, exclude type -(type+1)",
    "_comment": "                    for water (O:0, H:1) it can be",
    "_comment": "                    [0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0]",
    "fitting_neuron":   [240, 120, 60, 30, 10],

    "_comment": " traing controls",
    "systems":          ["../../data/Ru2Ru3/"],
    "set_prefix":       "set",
    "stop_batch":       1000000,
    "batch_size":       8,
    "start_lr":         0.001,
    "decay_steps":      5000,
    "decay_rate":       0.95,

    "start_pref_e":     0.02,
    "limit_pref_e":     8,
    "start_pref_f":     0,
    "limit_pref_f":     0,
    "start_pref_v":     0,
    "limit_pref_v":     0,

    "seed":             1,

    "_comment": " display and restart",
    "_comment": " frequencies counted in batch",
    "disp_file":        "lcurve.out",
    "disp_freq":        100,
    "numb_test":        10,
    "save_freq":        1000,
    "save_ckpt":        "model.ckpt",
    "load_ckpt":        "model.ckpt",
    "disp_training":    true,
    "time_training":    true,

    "_comment":         "that's all"
}
```

However, when I run this I get the warning message:
`warning: loc idx out of lower bound`
What causes this error, am I missing something obvious?

Thanks in advance,

Jim Boelrijk

**Comment by amcadmus** [comment, marked as answer]
deepmd-kit assumes the origin of the simulation box is (0, 0, 0).  For a cubic box, it is [0, L] x [0, L] x [0, L]. Of course, the box is not restricted to cubic and can be triclinic. 

If the positions of atoms are not in the simulation region, then the periodic boundary condition is applied and the atoms are wrapped back into the region by trying to +L or -L to the coordinates. However, this is tried only once, and if the atoms are too far away from the simulation region, the wrapping fails and the warning will be triggered. 

It is recommended that when preparing the data, all the coordinates of atoms are corrected wrapped to the simulation box with the origin (0, 0, 0). 

Please let me know if it fixes the warning.

**Comment by Jimbo994** [comment]
I followed up your advice of wrapping the data to the origin (0,0,0) and the error is gone.

Thanks for the help!

**Comment by feng-phoenix-wang** [comment]
Hi, I have a similar warning: loc idx out of upper bound, and I have already wrapped the coordinations. I also check coord.npy files. There is no number larger than L or less than 0. But this warning still occurs.
